runtime/common/src/main/kotlin/kotlinx/serialization/modules/SerialModule.kt

```diff
@@ -0,0 +1,97 @@
+@file:Suppress("UNCHECKED_CAST", "RedundantVisibilityModifier")
+
+package kotlinx.serialization.modules
+
+import kotlinx.serialization.*
+import kotlin.reflect.KClass
+
+/**
+ * Serial module is a collection of serializers used by [ContextSerializer] and [PolymorphicSerializer]
+ * to override or provide serializers at the runtime, whereas at the compile-time they provided by the
+ * serialization plugin.
+ *
+ * It can be regarded as a map where serializers are found using statically known KClasses.
+ *
+ * To allow a runtime resolving of serializers and usage of the particular SerialModule,
+ * one of the special annotations must be used.
+ *
+ * Typically, [SerialModule] would be passed as a constructor argument of [SerialFormat].
+ *
+ * @see ContextualSerialization
+ * @see Polymorphic
+ */
+public interface SerialModule {
+
+    /**
+     * Returns a dependent serializer associated with a given [kclass].
+     *
+     * This method is used in context-sensitive operations
+     * on a property marked with [ContextualSerialization], by a [ContextSerializer]
+     */
+    public operator fun <T: Any> get(kclass: KClass<T>): KSerializer<T>?
+
+    /**
+     * Returns serializer registered for polymorphic serialization of an [obj]'s class in a scope of [basePolyType].
+     *
+     * This method is used inside a [PolymorphicSerializer] when statically known class of a property marked with [Polymorphic]
+     * is [basePolyType], and the actual object in this property is [obj].
+     */
+    public fun <T : Any> resolveFromBase(basePolyType: KClass<T>, obj: T): KSerializer<out T>?
```

Comment:
[TBD] Why not base poly class?
We now have both "kclass" and "type" terminology in one API

Reply:
hm, we can discuss this on DM.